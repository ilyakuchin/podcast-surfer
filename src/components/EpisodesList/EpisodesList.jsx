import React, { useState } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { List, Image, Pagination } from 'semantic-ui-react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { setCurrentEpisode } from '../../redux/actions/CurrentEpisode/currentEpisode';

const Episodes = styled.ul`
  margin: 0;
  padding: 0;
  max-width: 700px;
  grid-area: episodes;
  list-style-type: none;
  display: grid;
  justify-items: center;
  align-items: center;
`;

export function EpisodesList({ episodes, setCurrentEpisodeConnect }) {
  const [currentPage, setCurrentPage] = useState(1);
  return (
    <Episodes>
      <Pagination
        defaultActivePage={1}
        totalPages={Math.round(episodes.length / 10)}
        onPageChange={(e, { activePage }) => {
          setCurrentPage(activePage);
        }}
      />
      <List verticalAlign='middle'>
        {episodes
          .slice((currentPage - 1) * 10, (currentPage - 1) * 10 + 10)
          .map(({ id, name, description, imageUrl, audioUrl }) => (
            <List.Item key={id}>
              <Image
                style={{
                  width: '200px',
                  height: '200px',
                  objectFit: 'scale-down'
                }}
                src={imageUrl}
              />
              <List.Content>
                <Link
                  onClick={() =>
                    setCurrentEpisodeConnect(
                      name,
                      description,
                      imageUrl,
                      audioUrl
                    )
                  }
                  to='/episode-player'
                >
                  {name}
                </Link>
              </List.Content>
            </List.Item>
          ))}
      </List>
    </Episodes>
  );
}

const mapStateToProps = state => {
  return {
    episodes: state.currentPodcast.episodes
  };
};

const dispatchToProps = dispatch => {
  return {
    setCurrentEpisodeConnect: (name, description, imageUrl, audioUrl) =>
      dispatch(setCurrentEpisode(name, description, imageUrl, audioUrl))
  };
};

EpisodesList.propTypes = {
  episodes: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      name: PropTypes.string,
      imageUrl: PropTypes.string
    })
  ).isRequired,
  setCurrentEpisodeConnect: PropTypes.func.isRequired
};

const ConnectedEpisodesList = connect(
  mapStateToProps,
  dispatchToProps
)(EpisodesList);
export default ConnectedEpisodesList;
