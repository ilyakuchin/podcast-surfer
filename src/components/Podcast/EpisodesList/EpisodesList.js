import React from "react";
import Episode from "../Episode/Episode";
import styled from "styled-components";

const Episodes = styled.ul`
  margin: 0;
  padding: 0;
  max-width: 700px;
  grid-area: episodes;
  list-style-type: none;
  display: grid;
  justify-items: center;
  align-items: center;
`;

export default function EpisodesList(props) {
  return (
    <Episodes>
      {props.episodes.map(item => (
        <Episode
          key={item.id}
          name={item.name}
          description={item.description}
          image={item.image}
          audio={item.audio}
        ></Episode>
      ))}
    </Episodes>
  );
}
