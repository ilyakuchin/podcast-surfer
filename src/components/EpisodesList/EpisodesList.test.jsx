import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure } from 'enzyme';
import toJson from 'enzyme-to-json';
import { EpisodesList } from './EpisodesList';

configure({ adapter: new Adapter() });

describe('Test EpisodesList Component', () => {
  test('should render correctly', () => {
    const episodes = [
      {
        id: '1',
        name: 'name1',
        description: 'description1',
        image: 'image1',
        audio: 'audio1'
      },
      {
        id: '2',
        name: 'name2',
        description: 'description2',
        image: 'image2',
        audio: 'audio2'
      },
      {
        id: '3',
        name: 'name3',
        description: 'description3',
        image: 'image3',
        audio: 'audio3'
      }
    ];
    expect(
      toJson(
        shallow(
          <EpisodesList
            rss='rss'
            episodes={episodes}
            setCurrentEpisodeConnect={jest.fn}
          />
        )
      )
    ).toMatchSnapshot();
  });

  test('should call fetchEpisodeConnect function on Link click', () => {
    const episodes = [
      {
        id: '1',
        name: 'name1',
        description: 'description1',
        image: 'image1',
        audio: 'audio1'
      },
      {
        id: '2',
        name: 'name2',
        description: 'description2',
        image: 'image2',
        audio: 'audio2'
      },
      {
        id: '3',
        name: 'name3',
        description: 'description3',
        image: 'image3',
        audio: 'audio3'
      }
    ];

    const fn = jest.fn();

    const wrapper = shallow(
      <EpisodesList
        rss='rss'
        episodes={episodes}
        setCurrentEpisodeConnect={fn}
      />
    );

    wrapper
      .find('Link')
      .first()
      .simulate('click');

    expect(fn.mock.calls.length).toEqual(1);
  });
});
