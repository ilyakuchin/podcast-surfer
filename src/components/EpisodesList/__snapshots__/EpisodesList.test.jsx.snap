// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test EpisodesList Component should render correctly 1`] = `
<styled.ul>
  <List
    verticalAlign="middle"
  >
    <ListItem
      key="1"
    >
      <Image
        as="img"
        style={
          Object {
            "height": "200px",
            "objectFit": "scale-down",
            "width": "200px",
          }
        }
        ui={true}
      />
      <ListContent>
        <Link
          onClick={[Function]}
          to="/episode-player"
        >
          name1
        </Link>
      </ListContent>
    </ListItem>
    <ListItem
      key="2"
    >
      <Image
        as="img"
        style={
          Object {
            "height": "200px",
            "objectFit": "scale-down",
            "width": "200px",
          }
        }
        ui={true}
      />
      <ListContent>
        <Link
          onClick={[Function]}
          to="/episode-player"
        >
          name2
        </Link>
      </ListContent>
    </ListItem>
    <ListItem
      key="3"
    >
      <Image
        as="img"
        style={
          Object {
            "height": "200px",
            "objectFit": "scale-down",
            "width": "200px",
          }
        }
        ui={true}
      />
      <ListContent>
        <Link
          onClick={[Function]}
          to="/episode-player"
        >
          name3
        </Link>
      </ListContent>
    </ListItem>
  </List>
</styled.ul>
`;
